qringbuf.go

```diff
@@ -255,16 +483,59 @@ func (qrb *QuantizedRingBuffer) signalCond(c chan<- struct{}) {
 	}
 }
 
+// Buffered returns the current amount of data already read from the underlying
+// reader, but not yet served via NextRegion(…). It is primarily useful for
+// informative error messages:
+//
+//  if err == io.ErrUnexpectedEOF {
+//    return fmt.Errorf(
+//      "unexpected end of stream after %d bytes (stream expected to be %d bytes long)",
+//      totalProcessedSoFar + int64( qrb.Buffered() ),
+//      expectedStreamLengthPassedToStartFill,
+//    )
+//  }
+func (qrb *QuantizedRingBuffer) Buffered() int {
+	defer qrb.Unlock()
+	qrb.Lock()
+	return qrb.cPos - (qrb.ePos + qrb.curRegionSize)
+}
+
+// NextRegion returns a *Region object representing a portion of the underlying
+// stream. One can explicitly request overlapping *Region's by supplying the
+// number of bytes to "step back" (use 0 for "just give me what's next"). This
```

Comment:
I expect seeking backwards will be common, right? In that case, how about re-defining `NextRegion` as `NextRegion(offset)` where `offset` seeks _forward_ from the start of the previous region. Calling this with `0` would just return the previously read region. That should lead to fewer arithmetic errors.

Reply:
We need to chat about this more...

At present the panic of "you gave me back everything I gave you" has saved me from countless endless-loops. Having to keep track of "what did I eat" and advancing your own "view" by hand seems... hrm.

As a general rule seeking back a super-short amount is common ( due to your last record having been cut ). Re-reading the entire buffer though is outright a bug.